### PR TITLE
Group entities by physical device

### PR DESCRIPTION
**Pull request by iMicknl**
Functionality from #328 by @vlebourl

### CODE REVIEW

**Comment by iMicknl** [comment]
@vlebourl could you give this a try? I don't have any devices that have multiple entities...

File: custom_components/tahoma/tahoma_device.py
Comment (iMicknl) [inline custom_components/tahoma/tahoma_device.py:70]: Those details are added as device information and in the debug log. Since last versions the attributes are added to the front-end, thus this is cluttering the view.

File: custom_components/tahoma/sensor.py
Comment (iMicknl) [inline custom_components/tahoma/sensor.py:29]: We use other code that is only compatible with the latest core versions.

File: custom_components/tahoma/__init__.py
Comment (iMicknl) [inline custom_components/tahoma/__init__.py:150]: TaHoma is already in the log, before the message. This applies to all changes here.

**Comment by vlebourl** [comment]
> @vlebourl could you give this a try? I don't have any devices that have multiple entities...

I'll try that tomorrow! But looks good to me.

**Comment by vlebourl** [comment]
that works for me!
image

**Review by vlebourl — DISMISSED** [review]

**Comment by iMicknl** [comment]
@vlebourl I see that I need to copy the device info as well, looks like it is now displaying the last one.

**Comment by vlebourl** [comment]
> @vlebourl I see that I need to copy the device info as well, looks like it is now displaying the last one.

ohyeah... missed that one

**Comment by vlebourl** [comment]
> > @vlebourl I see that I need to copy the device info as well, looks like it is now displaying the last one.
> 
> ohyeah... missed that one

That's why somewhere I retrieve the `base_url#1` label to get the device info.

**Comment by iMicknl** [comment]
> > > @vlebourl I see that I need to copy the device info as well, looks like it is now displaying the last one.
> > 
> > 
> > ohyeah... missed that one
> 
> That's why somewhere I retrieve the `base_url#1` label to get the device info.

Did this work in your previous PR?

**Comment by vlebourl** [comment]
hmm scratch that, that's actually the `model` part in the return of `tahoma_device` `device_info`

**Comment by iMicknl** [comment]
@vlebourl we can pull model and firmware from our parent device (#1) as well. I am not able to test it, but what if we just leave it empty for devices not being #1? Could you try that if I add it to this pr?

**Comment by vlebourl** [comment]
only `#1` info shows up now: 
image
that's fine IMO

**Comment by vlebourl** [comment]
but firmware is wrong...

**Comment by iMicknl** [comment]
@vlebourl could you perhaps give the latest version a try? If this doesn't work, I am not sure how to solve it :-). It is hard to fix without having the device present.

**Comment by iMicknl** [comment]
While removing old branches, I came across https://github.com/iMicknl/ha-tahoma/pull/152. Was this simple approach not good enough? :D

File: custom_components/tahoma/tahoma_device.py
Comment (iMicknl) [inline custom_components/tahoma/tahoma_device.py:118]:
```suggestion
                entity_registry = await self.hass.helpers.entity_registry.async_get_registry()
```

**Comment by tetienne** [comment]
@iMicknl If I remember well, it was because the name of the device was not correct. But indeed, the amount of modification between this PR and mine are not the same at all :)
@lebourl how dit you solve this: https://github.com/iMicknl/ha-tahoma/pull/152/files#r455789429 ?

**Comment by iMicknl** [comment]
> @iMicknl If I remember well, it was because the name of the device was not correct. But indeed, the amount of modification between this PR and mine are not the same at all :)

Ahh, that is indeed solved in this PR. The amount of code is indeed more, but it is not that much. Most of the code is around retrieving the entity id / name of the original device.

File: custom_components/tahoma/tahoma_device.py
Comment (tetienne) [inline custom_components/tahoma/tahoma_device.py:145]: Can you try to split the logic in several methods? That's not easy to understand at the first look.

File: custom_components/tahoma/tahoma_device.py
Comment (tetienne) [inline custom_components/tahoma/tahoma_device.py:134]: Why not return the other information? It introduces weird behavior?

File: custom_components/tahoma/tahoma_device.py
Comment (tetienne) [inline custom_components/tahoma/tahoma_device.py:109]: From what I see, you don't need this. You never use it. The next time we see it, you change its value [here]( https://github.com/iMicknl/ha-tahoma/pull/343/files#diff-14ad169bb8b77afd973fb894391dd4b0e629c3ebef801e129e618e09ce79a444R115) is enough.

Comment (iMicknl) [reply custom_components/tahoma/tahoma_device.py:109]: We need it for devices that have no subdevices. I will see if I can make the code more clear.

File: custom_components/tahoma/tahoma_device.py
Comment (iMicknl) [reply custom_components/tahoma/tahoma_device.py:134]: HA does use the device info from the last entity of a device, while we prefer to show information of the first device..

File: custom_components/tahoma/tahoma_device.py
Comment (tetienne) [reply custom_components/tahoma/tahoma_device.py:109]: At least perhaps rename it to something like `main_device_url`?

File: custom_components/tahoma/tahoma_device.py
Comment (tetienne) [reply custom_components/tahoma/tahoma_device.py]: The following information are not the same across the devices?
```
{
                "name": name,
                "manufacturer": manufacturer,
                "model": model,
                "sw_version": self.device.controllable_name,
            }
```

File: custom_components/tahoma/tahoma_device.py
Comment (iMicknl) [reply custom_components/tahoma/tahoma_device.py]: I pushed a new version which is totally different, thanks for your review. Hopefully it still works, unfortunately I am not able to test..

File: custom_components/tahoma/tahoma_device.py
Comment (iMicknl) [reply custom_components/tahoma/tahoma_device.py]: Not sure... Controllable name is different per Somfy entity for sure (io: xxx).

**Comment by iMicknl** [comment]
@vlebourl could you perhaps give this PR a try? I am not able to test it and would be good to check before merge.

File: custom_components/tahoma/tahoma_device.py
Comment (vlebourl) [inline custom_components/tahoma/tahoma_device.py:120]: Device with `#1` in the url should also return the base url to be merged with subsequent subdevices.

```suggestion
            "identifiers": {(DOMAIN, self.get_base_device_url())},
```

following, `get_base_device_url()` should be modified as well, see following comment.

File: custom_components/tahoma/tahoma_device.py
Comment (vlebourl) [inline custom_components/tahoma/tahoma_device.py:180]: following previous comment:
```suggestion
        device_url, _ = (
            self.device_url.split("#")
            if "#" in self.device_url
            else (self.device_url, "")
        )
```

**Review by vlebourl — CHANGES_REQUESTED** [review]

**Comment by vlebourl** [comment]
Never mind previous comments, I propose the following solution, better IMO

File: custom_components/tahoma/tahoma_device.py
Comment (vlebourl) [inline custom_components/tahoma/tahoma_device.py:113]:
```suggestion
                "identifiers": {(DOMAIN, self. device_base_url)},
```

**Review by vlebourl — CHANGES_REQUESTED** [review]

File: custom_components/tahoma/tahoma_device.py
Comment (iMicknl) [reply custom_components/tahoma/tahoma_device.py:180]: Thanks!

I will hold off a bit more on this PR, since it needs to take into account scenario's where deviceUrl#3 is supported, but where deviceUrl#1 is not supported (climate for example). In this case it should have some kind of a fallback.

**Review by vlebourl — APPROVED** [review]

**Pull request merged**
